File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var calcMid = function(left, right) {
    return Math.floor((left + right) / 2)
}

var searchMatrix = function(matrix, target) {
    const n = matrix[0].length
    const m = matrix.length
    
    let left = 0
    let right = m - 1
    let boundary = right
    let mid = null
    
    while (left <= right) {
        mid = calcMid(left, right)
        
        if (matrix[mid][0] === target) return true
        if (matrix[mid][0] < target) {
            boundary = mid
            left = mid + 1
        }
        else right = mid - 1
    }
    
    const y = boundary
    console.log(y)
    
    left = 0
    right = n - 1
    
    while (left <= right) {
        mid = calcMid(left, right)
        
        if (matrix[y][mid] === target) return true
        if (matrix[y][mid] > target) right = mid - 1
        else left = mid + 1
    }
    
    return false
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        
        l, r = 0, m - 1
        row = None
        while l <= r:
            mid = (l + r) // 2
            if target > matrix[mid][n - 1]:
                l = mid + 1
            elif target < matrix[mid][0]:
                r = mid - 1
            else:
                row = mid
                break
        if row == None: return False
        
        l, r = 0, n - 1
        while l <= r:
            mid = (l + r) // 2
            if target > matrix[row][mid]:
                l = mid + 1
            elif target < matrix[row][mid]:
                r = mid - 1
            else:
                return True
        return False
        
            
        